Ticket30920

JMGuisadoG opened the pull request:
[Trac link](https://trac.torproject.org/projects/tor/ticket/30920)

Nowrap multiplication function has its own commit. That commit has no change file since I was not sure if it needed a changelog for users. If needed I can amend that commit.

JMGuisadoG commented on src/lib/confmgt/unitparse.c:
```diff
@@ -145,7 +146,14 @@ config_parse_units(const char *val, const unit_table_t *u, int *ok)
-        v *= u->multiplier;
+        v = tor_mul_u64_nowrap(v, u->multiplier);
+
+      if ( v > SSIZE_MAX ) {
```
Would it be okay to check for `v > UINT64_MAX` instead of `SSIZE_MAX` ?

[Trac comment](https://trac.torproject.org/projects/tor/ticket/30920#comment:13) regarding using SSIZE_MAX instead of UINT64_MAX. Probably the difference between SSIZE_MAX in 64-bit machines and 32-bit machines are causing some tests to fail.

teor2345 commented on src/lib/confmgt/unitparse.c:
```diff
@@ -142,10 +143,26 @@ config_parse_units(const char *val, const unit_table_t *u, int *ok)
+      if (use_float) {
+        d = u->multiplier * d;
+
+        if ( d > INT64_MAX ) {
```
Let's make sure we're doing a floating point comparison here, using:
```suggestion
        if ( d > (double)INT64_MAX ) {
```

teor2345 replied:
See my updated suggestion, where I do both the integer and floating point checks.

JMGuisadoG replied:
Thank you! This helped learn a bit more about safety comparison.

I'll update with your suggestion when the tests are ok.

teor2345 commented on src/test/test_confparse.c:
```diff
@@ -906,11 +906,8 @@ test_confparse_unitparse(void *args)
   /* u64 overflow */
-  /* XXXX our implementation does not currently detect this. See bug 30920. */
-  /*
   tt_u64_op(config_parse_memunit("20000000 TB", &ok), OP_EQ, 0);
```
We'll also need tests for the floating-point case.
It would be great to have a test that fails the float comparison, and then one that passes the float comparison, but fails the int comparison.

JMGuisadoG replied:
I've never dealt with float operations in such a context so I had to read about it a bit before. I'm pasting here what I found (by testing in a separate c file) to be a case for each failing comparison before submitting.

In this case the double comparison will fail
```
 // Fails double comparison
 tt_u64_op(config_parse_memunit("15000000.5 TB", &ok), OP_EQ, 0);
 tt_assert(!ok);
```

This case is a valid float number, (`INT64_MAX` - 0.5) but its integer representation is greater than INT64_MAX.
```
 // Fails int comparison
 tt_u64_op(config_parse_memunit("9223372036854775806.5 B", &ok), OP_EQ, 0);
 tt_assert(!ok);
```

JMGuisadoG commented on src/test/test_confparse.c:
```diff
@@ -906,11 +906,14 @@ test_confparse_unitparse(void *args)
   tt_assert(!ok);
-  */
+  // Fails double comparison
+  tt_u64_op(config_parse_memunit("15000000.5 TB", &ok), OP_EQ, 0);
```
This test fails the double check as the float representing 15000000.5 TB is greater than `(double) INT64_MAX`

teor2345 replied:
Can you put this information in the code?

JMGuisadoG commented on src/lib/confmgt/unitparse.c:
```diff
@@ -142,10 +143,26 @@ config_parse_units(const char *val, const unit_table_t *u, int *ok)
+      if (use_float) {
+        d = u->multiplier * d;
+
+        if ( d > (double)INT64_MAX || (uint64_t)d > INT64_MAX ) {
```
Thank you for noting this Nick. Do you think a comment stating this would be necessary or would just bloat the code?

teor2345 replied:
We want another test
```suggestion
        if ( d > (double)INT64_MAX || d < 0 || (uint64_t)d > INT64_MAX ) {
```

teor2345 replied:
But the error message is different, so it should be a separate if block.

JMGuisadoG replied:
> But the error message is different, so it should be a separate if block.

Would it be ok to warn about "Negative or too big" in the same message so this way there's no need to add another if block?

teor2345 replied:
if blocks are cheap. But people's time is expensive, and dealing with confusing error messages takes time.

So just add a different message. The users will thank you :-)

JMGuisadoG commented on src/lib/confmgt/unitparse.c:
```diff
@@ -142,10 +143,26 @@ config_parse_units(const char *val, const unit_table_t *u, int *ok)
-        v = (uint64_t)(u->multiplier * d);
-      else
-        v *= u->multiplier;
+      if (use_float) {
```
> We should update the documentation of this function to note its behavior on error.

I'm understanding you mean `config_parse_units`, right?

By updating documentation would a comment above the function suffice,? (Like the one you've commented before about the new `tor_mul_u64_nowrap`)

JMGuisadoG commented on src/lib/intmath/muldiv.c:
```diff
@@ -69,6 +69,18 @@ gcd64(uint64_t a, uint64_t b)
   return a;
 }
 
+uint64_t
```
I could try giving it some documentation and accept any suggestions afterwards, which are likely given that I'm new to Tor contributions.

I can rewrite the commit for this function; or maybe, squashing this contribution into a single one is preferred?

teor2345 commented on src/test/test_confparse.c:
```diff
@@ -906,11 +906,14 @@ test_confparse_unitparse(void *args)
+  tt_u64_op(config_parse_memunit("15000000.5 TB", &ok), OP_EQ, 0);
+  tt_assert(!ok);
+  // Fails int comparison
+  tt_u64_op(config_parse_memunit("8388608.1 TB", &ok), OP_EQ, 0);
```
Can you put this information in the code as a comment?

JMGuisadoG commented:
I've updated the commits with your suggestions.

- Added documentation to the new nowrap mul function
- Added comments to the unitparse tests explaining each float representation case
- Added `d < 0` if-block and warn message in unitparse.c
- Added `d > 0` test inside float comparison if-block to avoid some compiler warnings in unitparse.c

teor2345 commented on src/lib/confmgt/unitparse.c:
```diff
@@ -142,10 +144,33 @@ config_parse_units(const char *val, const unit_table_t *u, int *ok)
+
+        if (d < 0) {
+          log_warn(LD_CONFIG, "Negative value arised when parsing %s %s",
+                   val, u->unit);
```
We also need to fail on negative values:
```suggestion
                   val, u->unit);
          *ok = 0;
           goto done;  
```

JMGuisadoG replied:
Ouch! Sorry for missing this.

I will also add a negative double test so everything is covered.
